api/src/main/java/io/grpc/TlsChannelCredentials.java

```diff
@@ -119,6 +242,126 @@ public Builder requireFakeFeature() {
       return this;
     }
 
+    /**
+     * Use the provided certificate chain and private key as the client's identity. Generally they
+     * should be PEM-encoded and the key is an unencrypted PKCS#8 key (file headers have "BEGIN
+     * CERTIFICATE" and "BEGIN PRIVATE KEY").
+     */
+    public Builder keyManager(File certChain, File privateKey) throws IOException {
+      return keyManager(certChain, privateKey, null);
+    }
+
+    /**
+     * Use the provided certificate chain and possibly-encrypted private key as the client's
+     * identity. Generally they should be PEM-encoded and the key is a PKCS#8 key. If the private
+     * key is unencrypted, then password must be {@code null}.
+     */
+    public Builder keyManager(File certChain, File privateKey, String privateKeyPassword)
+        throws IOException {
+      InputStream certChainIs = new FileInputStream(certChain);
+      try {
+        InputStream privateKeyIs = new FileInputStream(privateKey);
+        try {
+          return keyManager(certChainIs, privateKeyIs, privateKeyPassword);
+        } finally {
+          privateKeyIs.close();
+        }
+      } finally {
+        certChainIs.close();
+      }
+    }
+
+    /**
+     * Use the provided certificate chain and private key as the client's identity. Generally they
+     * should be PEM-encoded and the key is an unencrypted PKCS#8 key (file headers have "BEGIN
+     * CERTIFICATE" and "BEGIN PRIVATE KEY").
+     */
+    public Builder keyManager(InputStream certChain, InputStream privateKey) throws IOException {
+      return keyManager(certChain, privateKey, null);
+    }
+
+    /**
+     * Use the provided certificate chain and possibly-encrypted private key as the client's
+     * identity. Generally they should be PEM-encoded and the key is a PKCS#8 key. If the private
+     * key is unencrypted, then password must be {@code null}.
+     */
+    public Builder keyManager(
+        InputStream certChain, InputStream privateKey, String privateKeyPassword)
+        throws IOException {
+      byte[] certChainBytes = ByteStreams.toByteArray(certChain);
+      byte[] privateKeyBytes = ByteStreams.toByteArray(privateKey);
+      clearKeyManagers();
+      this.certificateChain = certChainBytes;
+      this.privateKey = privateKeyBytes;
+      this.privateKeyPassword = privateKeyPassword;
+      return this;
+    }
+
+    /**
+     * Have the provided key manager select the client's identity. Although multiple are allowed,
+     * only the first instance implementing a particular interface is used. So generally there will
+     * just be a single entry and it implements {@link javax.net.ssl.X509KeyManager}.
+     */
+    public Builder keyManager(KeyManager... keyManagers) {
```

Comment:
This is quite similar to `SslContextBuilder` and the intent might be to provide similar functionality.

- how about a method to accept a `KeyManagerFactory` like the SslContextBuilder? I see that you are anyway wrapping this list in a `FixedKeyManagerFactory`
- then just use a `KeyManagerFactory` inside and use a `KeyManagerFactoryWrapper` or `FixedKeyManagerFactory` to support these other KeyManager setters.

Reply:
I'm making the grpc-netty implementation more complicated to workaround a useless Netty-ism (that should really be fixed) while making the user-facing API easier to use.

Java [accepts the managers](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html#init-javax.net.ssl.KeyManager:A-javax.net.ssl.TrustManager:A-java.security.SecureRandom-), not the factories. The factories are actually a giant pain, because they require a Provider and an algorithm and that stuff doesn't actually matter to any consumer of the object.

KeyManager and TrustManager are much easier for users to implement their own. For example, for the "advanced TLS" APIs, they will provide a custom KeyManager and TrustManager implementation. I don't want to _force_ the user to convert those to a Factory.

Netty uses the Factories in its API, and it seems clearly a mistake. The `Fixed*Factory`s are a workaround, and they do weird things. I should really be changing Netty to use Manager lists internally; I've checked and I only saw calls the `get*Managers()` method. That would remove the need for _anyone_ to use `*ManagerFactoryWrapper` that is in Netty or the `Fixed*ManagerFactory` that I add here.